File: samples/Test/Classes/Control/PickerViewTest.cpp

#include "PickerViewTest.h"

#define _T(x) L##x


PickerViewTest::PickerViewTest()
:m_iIndex(0)
{
    this->setTitle("CAPickerView");
    
    m_vCitys =
    {
        _T("北京市"),
        _T("天津市"),
        _T("上海市"),
        _T("重庆市"),
        _T("河北省"),
        _T("山西省"),
        _T("辽宁省"),
        _T("吉林省"),
        _T("黑龙江省"),
        _T("江苏省"),
        _T("浙江省"),
        _T("安徽省"),
        _T("福建省"),
        _T("江西省"),
        _T("山东省"),
        _T("河南省"),
        _T("湖北省"),
        _T("湖南省"),
        _T("广东省"),
        _T("海南省"),
        _T("四川省"),
        _T("贵州省"),
        _T("云南省"),
        _T("陕西省"),
        _T("甘肃省"),
        _T("青海省"),
        _T("台湾省"),
        _T("广西壮族自治区"),
        _T("内蒙古自治区"),
        _T("西藏自治区"),
        _T("宁夏回族自治区"),
        _T("新疆维吾尔自治区"),
        _T("香港特别行政区"),
        _T("澳门特别行政区"),
    };
}

PickerViewTest::~PickerViewTest()
{
}



void PickerViewTest::viewDidLoad()
{
    if (m_iIndex == 0)
    {
        auto label = CALabel::createWithLayout(DLayout(DHorizontalLayoutFill, DVerticalLayout_T_H(100, 40)));
        label->setText(UTF8("北京市"));
        label->setColor(CAColor4B::BLACK);
        label->setFontSize(28);
        label->setTextAlignment(CATextAlignment::Center);
        label->setVerticalTextAlignmet(CAVerticalTextAlignment::Center);
        this->getView()->addSubview(label);
        
        auto pickerView = CAPickerView::createWithLayout(DLayout(DHorizontalLayout_NW_C(0.8, 0.5), DVerticalLayout_T_H(200, 400)));
        pickerView->setFontSizeNormal(40);
        pickerView->setFontSizeSelected(40);
        this->getView()->addSubview(pickerView);
        pickerView->setBackgroundImage(CAImage::createWithColor4B(CAColor4B::YELLOW));
        
        pickerView->onTitleForRow([=](unsigned int row, unsigned int component){

            return unicode_to_utf8(m_vCitys.at(row));
        });
        
        pickerView->onHeightForComponent([=](unsigned int component){
            
            return 80;
        });
        
        pickerView->onWidthForComponent([=](unsigned int component){
            
            return this->getView()->getBounds().size.width * 0.8f;
        });
        
        pickerView->onNumberOfRowsInComponent([=](unsigned int component)
        {
            return m_vCitys.size();
        });
    
        pickerView->onViewForSelected([=](unsigned int component, DSize size){
            
            return CAView::createWithColor(CAColor4B::MAGENTA);
        });
        
        pickerView->onDidSelectRow([=](unsigned int row, unsigned int component){
            
            label->setText(unicode_to_utf8(m_vCitys[row]));
        });
       
    }
    
    else if(m_iIndex == 1)
    {
        CADatePickerView* dpv = CADatePickerView::createWithLayout(DLayout(DHorizontalLayout_NW_C(0.8, 0.5), DVerticalLayout_T_H(200, 400)),CADatePickerView::Mode::Date);
        dpv->setBackgroundImage(CAImage::createWithColor4B(CAColor4B::YELLOW));
        dpv->setBackgroundImageForSelected(CAImage::createWithColor4B(CAColor4B::GREEN));
        this->getView()->addSubview(dpv);
        
        //dpv->setDate(2004, 2, 16, false);
        dpv->onSelectRow(CALLBACK_BIND_1(PickerViewTest::dataPickerViewdidSelectRow, this));
        
    }
    else if(m_iIndex == 2)
    {
        CADatePickerView* dpv3 = CADatePickerView::createWithLayout(DLayout(DHorizontalLayout_NW_C(0.8, 0.5), DVerticalLayout_T_H(200, 400)),CADatePickerView::Mode::DateAndTime);
        dpv3->setBackgroundImage(CAImage::createWithColor4B(CAColor4B::YELLOW));
        dpv3->setBackgroundImageForSelected(CAImage::createWithColor4B(CAColor4B::GREEN));
        this->getView()->addSubview(dpv3);
        
        dpv3->onSelectRow(CALLBACK_BIND_1(PickerViewTest::dataPickerViewdidSelectRow, this));
    
    }
    else if(m_iIndex == 3)
    {
        CADatePickerView* dpv4 = CADatePickerView::createWithLayout(DLayout(DHorizontalLayout_NW_C(0.8, 0.5), DVerticalLayout_T_H(200, 400)),CADatePickerView::Mode::Time);
        dpv4->setBackgroundImage(CAImage::createWithColor4B(CAColor4B::YELLOW));
        dpv4->setBackgroundImageForSelected(CAImage::createWithColor4B(CAColor4B::GREEN));
        this->getView()->addSubview(dpv4);
        
        dpv4->onSelectRow(CALLBACK_BIND_1(PickerViewTest::dataPickerViewdidSelectRow, this));
    }
    else
    {
        CADatePickerView* dpv5 = CADatePickerView::createWithLayout(DLayout(DHorizontalLayout_NW_C(0.8, 0.5), DVerticalLayout_T_H(200, 400)),CADatePickerView::Mode::CountDownTimer);
        dpv5->setBackgroundImage(CAImage::createWithColor4B(CAColor4B::YELLOW));
        dpv5->setBackgroundImageForSelected(CAImage::createWithColor4B(CAColor4B::GREEN));
        this->getView()->addSubview(dpv5);
        
        dpv5->onSelectRow(CALLBACK_BIND_1(PickerViewTest::dataPickerViewdidSelectRow, this));
    
    }
    
    auto btn = CAButton::createWithLayout(DLayout(DHorizontalLayout_NW_C(0.3,0.5), DVerticalLayout_B_H(20 ,50)), CAButton::Type::RoundedRect);
    btn->setTitleForState(CAControl::State::Normal, "Switch Next");
    btn->addTarget([&] {
        
        ssize_t index = m_iIndex + 1;
        if (index > 4)
        {
            index = 0;
        }
        PickerViewTest* viewController = PickerViewTest::create();
        viewController->setIndex(index);
        RootWindow::getInstance()->getRootNavigationController()->replaceViewController(viewController, false);
        
    }, CAButton::Event::TouchUpInSide);
    this->getView()->addSubview(btn);
}

void PickerViewTest::viewDidUnload()
{
    // Release any retained subviews of the main view.
    // e.g. self.myOutlet = nil;
}

void PickerViewTest::dataPickerViewdidSelectRow(const tm& tm)
{
    CCLog("didSelectRow tm: %d-%d-%d",tm.tm_year+1900,tm.tm_mon+1,tm.tm_mday);
}

File: samples/Test_JS/Resources/js/Control/PickerViewTest.js
/**
 * Created by crossApp on 16/8/3.
 */



var PickerViewTest = ca.CAViewController.extend({

    ctor: function () {
        this._super();

        this.cityArr = [
            "北京市",
            "天津市",
            "上海市",
            "重庆市",
            "河北省",
            "山西省",
            "辽宁省",
            "吉林省",
            "黑龙江省",
            "江苏省",
            "浙江省",
            "安徽省",
            "福建省",
            "江西省",
            "山东省",
            "河南省",
            "湖北省",
            "湖南省",
            "广东省",
            "海南省",
            "四川省",
            "贵州省",
            "云南省",
            "陕西省",
            "甘肃省",
            "青海省",
            "台湾省",
            "广西壮族自治区",
            "内蒙古自治区",
            "西藏自治区",
            "宁夏回族自治区",
            "新疆维吾尔自治区",
            "香港特别行政区",
            "澳门特别行政区"
        ];
    },

    viewDidLoad: function() {

        // this.city_value = ca.CALabel.createWithLayout(ca.DLayout.set(ca.DHorizontalLayout.fill(), ca.DVerticalLayout_T_H(100, 40)));
        // this.city_value.setText("北京市");
        // this.city_value.setColor(ca.CAColor4B.BLACK);
        // this.city_value.setFontSize(28);
        // this.city_value.setTextAlignment(ca.CATextAlignment.Center);
        // this.city_value.setVerticalTextAlignmet(ca.CAVerticalTextAlignment.Center);
        // this.getView().addSubview(this.city_value);
        //
        //
        // this.pickerView = ca.CAPickerView.createWithLayout(ca.DLayout.set(ca.DHorizontalLayout_W_C(480, 0.5), ca.DVerticalLayout_T_H(200, 400)));
        // this.pickerView.setFontSizeNormal(40);
        // this.pickerView.setFontSizeSelected(40);
        // this.pickerView.setFontColorNormal(ca.CAColor4B.BLACK);
        // this.pickerView.setFontColorSelected(ca.CAColor4B.BLACK);
        // this.pickerView.setBackgroundImage(ca.CAImage.createWithColor4B(ca.CAColor4B.YELLOW));
        // this.getView().addSubview(this.pickerView);
        //
        // this.pickerView.onDidSelectRow(this.didSelectRow.bind(this)) ;
        // this.pickerView.onNumberOfComponents(this.numberOfComponentsInPickerView.bind(this)) ;
        // this.pickerView.onNumberOfRowsInComponent(this.numberOfRowsInComponent.bind(this)) ;
        // this.pickerView.onWidthForComponent(this.widthForComponent.bind(this)) ;
        // this.pickerView.onHeightForComponent(this.rowHeightForComponent.bind(this)) ;
        // this.pickerView.onTitleForRow(this.titleForRow.bind(this)) ;
        // this.pickerView.onViewForSelected(this.viewForSelected.bind(this));
        //
        // this.pickerView.reloadAllComponents();

        var dpv3 = ca.CADatePickerView.createWithLayout(ca.DLayout.set(ca.DHorizontalLayout_NW_C(0.8, 0.5), ca.DVerticalLayout_T_H(200, 400)),ca.CADatePickerView.Mode.Date);
        dpv3.setBackgroundImage(ca.CAImage.createWithColor4B(ca.CAColor4B.YELLOW));
        dpv3.setBackgroundImageForSelected(ca.CAImage.createWithColor4B(ca.CAColor4B.GREEN));
        this.getView().addSubview(dpv3);

        dpv3.onSelectRow(function (data) {

        });
    },
    didSelectRow: function ( row, component) {
        this.city_value.setText(this.cityArr[row]);
    },
    numberOfComponentsInPickerView: function () {
        return 1;
    },

    numberOfRowsInComponent: function ( component) {
        return this.cityArr.length;
    },
    widthForComponent: function ( component) {
        return 480;
    },

    rowHeightForComponent: function ( component) {
        return 80;
    },
    titleForRow: function ( row, component) {
        return this.cityArr[row];
    },
    viewForSelected: function (component, size) {
        return ca.CAView.createWithColor(ca.CAColor4B.GREEN);
    }
});
